#166272346 Allow users to follow other users

adex001 opened the pull request:
#### What does this PR do?
- Users can follow each other
#### Description of Task to be completed?
- Create a follow endpoint
- Create an unfollow endpoint
- Verify the endpoint and ensure that only valid users have access to that endpoint.
- write the follow functionality
- write the unfollow functionality
- make sure users cannot follow themselves
- make sure users cannot unfollow themselves
- make sure a user cannot follow the same user multiple times
- users cannot unfollow the same user more than one time


#### How should this be manually tested?
Navigate to the endpoint `{{HOST_NAME}}`/api/v1/users/:username/follow

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166272346](https://www.pivotaltracker.com/story/show/166272346)
#### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/29796551/59330887-96197f00-8cea-11e9-8a9b-bc08e8255cad.png)
### Questions:
N/A

adex001 committed:
feat-[166272346]: Users follow each other
create a follow endpoint
verify the endpoint to allow only valid users
write the follow functionality
make sure users cannot follow themselves
make sure a user cannot follow another user multiple times
update the routes to be version compliant

adex001 committed:
Merge branch 'develop' of https://github.com/andela/kingsmen-ah-backend into 166272346-ft-users-follow-each-other

adex001 committed:
feat-[166272346]: Users follow each other
update and write more tests

despeauxz commented on server/controllers/users.js:
```diff
@@ -150,6 +150,60 @@ class UserController {
+      if (followid === id) {
+        return res.status(400).json({
+          status: 400,
+          error: 'you cannot follow yourself',
```
Please use the `gloabl` error response

adex001 replied:
alright. Thanks.

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -150,6 +150,60 @@ class UserController {
+        });
+      }
+      if (followed) {
+        return res.status(201).json({
```
please take a look at the readme page to know what additional info to return

adex001 replied:
alright! I will;

adex001 replied:
There is no additional info to return. Moreover, I will continue the unfollow endpoint when I am able to charge my system.

adex001 committed:
feat-[166272346]: Users follow each other
update and write more test
pull from develop and resolve merge conflict
used faker to mock up datas
change test to use Bearer <token> as authorization

fidelisojeah commented on server/routes/api/index.js:
```diff
@@ -1,8 +1,11 @@
 
-apiRoutes.use('/', userRoutes);
+apiRoutes.use('/auth', userRoutes);
+apiRoutes.use('/users/:followid/follow', Token.verifyToken, UserController.follow);
```
shouldn't this be:
`apiRoutes.post` ?

However, 

- Create a profiles routes file

change this to:
`apiRoutes.use('/profile', Token.verifyToken, <user routes router>)`

and define the follow route there
`profileRoutes.post('/:username/follow', UserController.follow);`

adex001 replied:
Alright!

fidelisojeah commented on server/tests/follower.test.js:
```diff
@@ -0,0 +1,149 @@
+  return { username, email, password };
+};
+
+describe('TESTS TO CREATE USERS', () => {
```
why exactly are you testing creation of new users again?

adex001 replied:
I wanted to grab the token.

fidelisojeah commented on server/routes/api/index.js:
```diff
@@ -1,8 +1,11 @@
 import express from 'express';
 import userRoutes from './users';
+import Token from '../../helpers/Token';
```
Don't use complex relative paths (that need you going back .

This should have been 
`import Token from '@helpers/Token';`

adex001 replied:
Ok. Thanks

fidelisojeah commented on server/routes/api/index.js:
```diff
@@ -1,8 +1,11 @@
 
-apiRoutes.use('/', userRoutes);
+apiRoutes.use('/auth', userRoutes);
+apiRoutes.use('/users/:followid/follow', Token.authorize, UserController.follow);
```
??????????????????????????????????????????????????????????????????????????????????????????????????????

This should have been `apiRoutes.post ...`

However, 
create a new router file `profileRoutes`
change this to 
`apiRoutes.use('/profiles', Token.authorize, <new profile router>);`

Then in this new profile router define this as:

`profileRoutes.post('/:username/follow, UserController.follow);`

adex001 replied:
Thanks. I will implement this now.

fidelisojeah reviewed (changes requested):
Your ticket requires that users should be able to unfollow each other as well

adex001 committed:
feat-[166272346]: Users follow each other
update feedback from PR

adex001 committed:
feat-[166272346]: Users follow each other
update feedback from PR and update tests

adex001 committed:
feat-[166272346]: Users follow each other
update merge conflict by pulling from develop
update endpoint to fit in documentation
write unfollow functionality
write unit tests for unfollow endpoint

adex001 committed:
feat-[166272346]: Users follow each other
update merge conflict by pulling from develop
update endpoint to fit in documentation
write unfollow functionality
write unit tests for unfollow endpoint

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -90,12 +92,125 @@ class UserController {
+
+      const userToFollow = await User.findOne({
+        where: {
+          username,
```
```suggestion
          username,
          active: true
```

adex001 replied:
implemented!

fidelisojeah commented on server/routes/index.js:
```diff
@@ -0,0 +1,8 @@
+import express from 'express';
```
is any of these used?

adex001 replied:
yes. It's used for the API versioning.

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -90,12 +92,125 @@ class UserController {
+      if (followid === id) {
+        return Response.error(res, 400, 'you cannot follow yourself');
+      }
+      const [followed, created] = await Follower.findOrCreate({
```
This could be expressed some other way.

please look through the sequelize documentation

`user.addFollower(req.user)`
(user here is the user to be followed)

You may want to modify the `Follower` model.

change `followingId` to `userId` 

having `follower` and `following` makes very little sense 
either someone is being followed or someone is following someone

in our case, would be nice to have a user being followed (so a user has followers)

adex001 replied:
done eventually.

despeauxz reviewed (dismissed):
Nice Job

fidelisojeah commented on server/tests/follower.tests.js:
```diff
@@ -0,0 +1,210 @@
+
+describe('TESTS TO FOLLOW A USER', () => {
+  before(async () => {
+    const user1 = await models.User.create(userOne);
```
move user generation into the factory (and call the needed functions from here)

adex001 replied:
ok. implemented.

adex001 committed:
all

adex001 committed:
update feedback

adex001 committed:
Update the follow and unfollow using magic method

adex001 committed:
feat-[166272346]: Users follow each other
update feedback

fidelisojeah commented on server/controllers/users.js:
```diff
@@ -90,12 +92,130 @@ class UserController {
+      const { username } = req.params;
+      const { id } = req.decoded;
+
+      const me = await User.findByPk(id);
```
you could use `req.user`

adex001 replied:
implemented. Thanks.

adex001 committed:
feat-[166272346]: Users follow each other
updated feedback by using req.user in the follow controller
updated route index.js

adex001 committed:
feat-[166272346]: Users follow each other
resolve merge conflicts

adex001 committed:
feat-[166272346]: Users follow each other
add the follow endpoint to return a profile for  both follow anf unfolloe

adex001 committed:
feat-[166272346]: Users follow each other
add the create-follower file back to migrations

adex001 committed:
feat-[166272346]: Users follow each other
move the follow and unfollow controller from user controller to profile controller
update the follow and unfollow method to return profile of who followed

adex001 committed:
feat-[166272346]: Users follow each other
move the follow and unfollow controller from user controller to profile controller
update the follow and unfollow method to return profile of who followed

adex001 committed:
feat-[166272346]: Users follow each other
move the follow and unfollow controller from user controller to profile controller
update the follow and unfollow method to return profile of who followed

adex001 committed:
feat-[166272346]: Users follow each other
resolved merge conflicts

adex001 committed:
feat-[166272346]: Users follow each other
resolved merge conflicts

fidelisojeah commented on server/controllers/profile.js:
```diff
@@ -77,6 +77,91 @@ class ProfileController {
+      return res.status(200).json({
+        status: 200,
+        message: 'User unfollowed successfully',
+        payload: myProfile || [],
```
```suggestion
        payload: myProfile || {},
```

fidelisojeah commented on server/controllers/profile.js:
```diff
@@ -77,6 +77,91 @@ class ProfileController {
+      return res.status(201).json({
+        status: 201,
+        message: 'User followed successfully',
+        payload: myProfile || [],
```
```suggestion
        payload: myProfile || {},
```

adex001 replied:
just seeing this. Would implement.

adex001 committed:
feat-[166272346]: Users follow each other
add more tests for updateProfile when there is a validation error

adex001 committed:
feat-[166272346]: Users follow each other
return an object in the optional payload in both follow and unfollow endpoint
